File: app/src/main/java/com/mdgd/pokemon/ui/pokemon/PokemonDetailsViewModelFactory.java
package com.mdgd.pokemon.ui.pokemon;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.mdgd.pokemon.models.AppModule;

public class PokemonDetailsViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private final AppModule appComponent;

    public PokemonDetailsViewModelFactory(AppModule appComponent) {
        this.appComponent = appComponent;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass == PokemonDetailsViewModel.class) {
            return (T) new PokemonDetailsViewModel(appComponent.getPokemonsRepo());
        }
        return null;
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemon/PokemonDetailsViewModelFactory.kt
package com.mdgd.pokemon.ui.pokemon

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.mdgd.pokemon.models.AppModule

class PokemonDetailsViewModelFactory(private val appComponent: AppModule) : ViewModelProvider.NewInstanceFactory() {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return if (modelClass == PokemonDetailsViewModel::class.java) {
            PokemonDetailsViewModel(appComponent.getCache()) as T
        } else super.create(modelClass)
    }
}
